### DISCUSSION TITLE
Operations on bayered images

### DISCUSSION THREAD

**Discussion by ewelot** [libvips/libvips#2087] [Q&A]
Large, cooled one-shot-color CMOS cameras are gaining increased attention in astrophotography. The light is passing a Bayer matrix filter before reaching the sensor and therefore pixels in the resulting grey image represent different colors.

Several python wrappers for libraw do already exist (like [pyraw](https://github.com/letmaik/rawpy)) but they are designed to work with DSLR camera images and heavily rely on their metadata information as far as I know. Images created by the CMOS cameras I mentioned are not supported. Those are typically (single band) FITS images.

What I am looking for at the moment is some way of using libvips functionality for preprocessing of the bayered image, in particular
- extract pixels for each of the four Bayer matrix cells (e.g. RGGB) to create 4 separate grey sub-images (half-width and half-height with respect to the original image) representing different colors
- do some processing on individual colors (e.g. find/replace bad pixel)
- replace all pixels representing a given color in the original image by a processed sub-image

Whould you have any recommendation about how to achieve the extraction and replacement steps?
(BTW: I am currently using graphicsmagicks convert for the extraction like `gm convert <bayered_img> -roll +0+0 -filter point -resize 50% <single_color_img>` but I'd much prefer a pyvips solution)

**Comment by jcupitt** [comment]
Hi again,

You can do a quick subsample with eg. `image.subsample(2,2)`. It'll pick pixel (0,0), (2,0), (4,0), etc.

You can offset an image quickly with `embed`. So you can sample pixels (1,0), (3,0), (5,0), etc. with `image.embed(1, 0, image.width, image,height).subsample(2, 2)`.

I've fixed bad pixels by making a pixel map, then for each image computing a median and using ifthenelse to flip between the median image and the pixels, eg. `bad_pixel_map.ifthenelse(image.median(3), image)`. libvips is lazy, so it'll only compute the median in the area around the bad pixel.

When I've calibrated cameras, I've also modelled gain and offset for each sensor site. There can be quite a degree of variation (several percent) due to inconsistencies in the manufacturing process. I would take a lens-cap black for the offset calibration and do something like: `image = image - lens_cap_black + 7`. The seven is to make sure there's a slight positive offset (7 is 0.5, assuming a 16-bit image) so you don't clip off the negative noise floor. You'll obviously need to make sure you have the sensor in linear mode.

For gain, take a shot with the lens off and the bare sensor exposed. Correct it for offset, then for each image compute eg. `image = image * white.max() / white`.

Gain and offset will vary with sensor temperature, so be cautious.

You'll also need to correct for lens vignetting, aperture, etc.

**Comment by jcupitt** [comment]
Oh, the median ifthenelse thing will only work if your bad pixels are fairly spread out. A large cluster of them will cause problems. But I imagine bad pixels are pretty rare in the type of sensor you are using.

**Comment by jcupitt** [comment]
Sorry, one more thing, a little C for this first stage of calibration is easy and quite a bit quicker, if speed is important. You can certainly prototype in pyvips.

**Comment by ewelot** [comment]
Thanks for directing me to the `image.embed` and `image.subsample` functions. This is what I am looking for to extract individual Bayer cell colors into 4 separate half-size images. I can use those for the replacement of bad pixels, e.g. by some median filtering and the lazy processing of selected pixels only will certainly boost performance. Now I would need to rebuild the full size bayered image. Any idea on how to do this? The resulting bayered image is then converted to rgb using some debayering interpolation algorithm in the next step using an external program.

I fully agree with your calibration pipeline. It is exactly what I'm doing now: dark calibration (depending on gain, temperature and exposure time) and "flat field" calibration to account for vignetting, sensor pixel sensitivity variation etc. Moreover I need to assure a linear sensor response and of course a linear data processing. In fact, I'am interested in extended, very low surface brightness objects and the quality of calibrations is crucial. I do have a working pipeline already (without libvips) but would like to replace some slow processing steps by much faster libvips/pyvips operations.

**Comment by jcupitt** [comment]
You'd could use `zoom(2,2)` to quickly make four full-size images, and `case` to use an index image to select pixels from an array of source images. For the index image, you'd start with eg. `[[0, 1], [2, 3]]` as a 2x2 constant, then expand that as needed with `replicate`.

Something like (untested):

```python
images = [image.zoom(2,2) for image in [red, green1, blue, green2]]
index = pyvips.Image.new_from_array([[0, 1], [2, 3]]).replicate(green1.width + 1, green1.height + 1)
bayered = index.case(images)
```

**Comment by jcupitt** [comment]
Your pipeline sounds good. I would be tempted to do both a no-lens flatfield and a lens flatfield to separate pixel gain correction from lens vignetting but if you seldom change objectives perhaps it's more sensible to combine them.

**Comment by jcupitt** [comment]
Some sensors have a row of masked pixels down one edge (the chip is manufactued with a metal mask over those sites), it'd be worth checking if you have them available. They are very handy for correcting for temperature and exposure time in one go.

**Comment by ewelot** [comment]
Great info on how to combine the corrected sub-images and build up the corrected bayered image!

Yes, the 'overscan area' which is present in some sensor image data is a first approximation of its dark level. Obviously it is used in many cameras for this reason internally - even if it is not delivered in the raw data file to the user. The problem in astrophotography is the usually long exposure time which on most sensors creates a significant amp glow from heat generated by readout electronics. This is obviously more pronounced towards edge/edges or corner/corners and the only way to account for it is by taking full frame dark images.

I wonder if it would be of more general interest to implement a new function which does e.g. mean and/or median filtering based on an arbitrary kernel image (bi-level mask), somewhat similar to `vips_rank()` but using only 'white' pixels from a mask instead of all pixels within the filter window. This would certainly simplify the discussed replacement of bad pixels. I would also have other needs for such a filter method, e.g. to smooth out motion-blurred images, taking into account pixels perpendicular to the motion vector only. Do you think it is worth adding this funtionality in the foreseeable future?

**Comment by jcupitt** [comment]
Yes, that's a nice idea. You could just add an optional `mask` parameter to `rank` with the structuring element. It shouldn't be too hard, if you fancy a go.

**Comment by ewelot** [comment]
Shall I open a new issue tagged as 'feature request' for this extension of the rank function?

**Comment by ewelot** [comment]
By the way, this is the code I ended up with. Knowing the right functions it was pretty easy (just a few corrections to your hints) ...
```
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file("bayer.pgm")
badmask = pyvips.Image.new_from_file("badmask.png")

w = image.width
h = image.height
p00 = image.subsample(2, 2).median(3).zoom(2,2)
p10 = image.embed(-1, 0, w, h).subsample(2, 2).median(3).zoom(2,2)
p01 = image.embed(0, -1, w, h).subsample(2, 2).median(3).zoom(2,2)
p11 = image.embed(-1, -1, w, h).subsample(2, 2).median(3).zoom(2,2)

index = pyvips.Image.new_from_array([[0, 1], [2, 3]]).replicate(w/2, h/2)
median = index.case([p00, p10, p01, p11])

cleaned = (badmask > 0).ifthenelse(median, image)
cleaned.ppmsave("cleaned.pgm", strip=1)
```
